File: client/src/types/badge.ts
export interface Badge {
  id: string;
  badgeType: string;
  name: string;
  description: string;
  icon: string;
  color: string;
  earnedAt: string;
  progress: number;
  requiredProgress: number;
  isCompleted: boolean;
}

export interface UserProgress {
  totalQuizzesCompleted: number;
  perfectScores: number;
  totalScore: number;
  totalQuestionsAnswered: number;
  algorithmQuizCounts: Record<string, number>;
  currentStreak: number;
  longestStreak: number;
  badges: Badge[];
}

export interface QuizCompletion {
  algorithmType: string;
  score: number;
  totalQuestions: number;
  isPerfectScore: boolean;
  streak: number;
}

export interface BadgeNotification {
  badge: Badge;
  show: boolean;
} 

File: Api/Models/Badge.cs
using System;
using Google.Cloud.Firestore;

namespace Api.Models
{
    [FirestoreData]
    public class Badge
    {
        [FirestoreDocumentId]
        public string Id { get; set; }
        
        [FirestoreProperty]
        public string UserId { get; set; }
        
        [FirestoreProperty]
        public string BadgeType { get; set; }
        
        [FirestoreProperty]
        public string Name { get; set; }
        
        [FirestoreProperty]
        public string Description { get; set; }
        
        [FirestoreProperty]
        public string Icon { get; set; }
        
        [FirestoreProperty]
        public string Color { get; set; }
        
        [FirestoreProperty]
        public DateTime EarnedAt { get; set; }
        
        [FirestoreProperty]
        public int Progress { get; set; }
        
        [FirestoreProperty]
        public int RequiredProgress { get; set; }
        
        [FirestoreProperty]
        public bool IsCompleted { get; set; }
    }
} 